pint_xarray/checking.py

```diff
@@ -0,0 +1,186 @@
+    return_values : Any
+        Return values of the wrapped function, either a single value or a tuple of values. These will be given units
+        according to return_units.
+
+    Raises
+    ------
+    TypeError
+        If an argument or return value has a specified unit, but is not an xarray.DataArray or pint.Quantity.
+        Also thrown if any of the units are not a valid type, or if the number of arguments or return values does not
+        match the number of units specified.
+
+    Examples
+    --------
+
+    Decorating a function which takes one quantified input, but returns a non-data value (in this case a boolean).
+
+    >>> @expects("deg C")
+    ... def above_freezing(temp):
+    ...     return temp > 0
+
+
+    TODO: example where we check units of an optional weighted kwarg
+    """
+
+    # Check types of args_units, kwargs_units, and return_units
+    all_units = list(args_units) + list(kwargs_units.values())
+    if isinstance(return_units, list):
+        all_units = all_units + return_units
+    elif return_units:
+        all_units = all_units + [return_units]
+    for a in all_units:
+        if isinstance(a, dict):
+            for u in a.values():
+                _check_valid_unit_type(u)
+        else:
+            _check_valid_unit_type(a)
+
+    def _expects_decorator(func):
+        @functools.wraps(func)
+        def _unit_checking_wrapper(*args, **kwargs):
+
+            # check same number of arguments were passed as expected
+            if len(args) != len(args_units):
+                raise TypeError(
+                    f"{len(args)} arguments were passed, but {len(args_units)} arguments were expected"
+                )
+
+            converted_args = []
+            for arg, arg_unit in zip(args, args_units):
+                converted_arg = _check_or_convert_to_then_strip(arg, arg_unit)
+                converted_args.append(converted_arg)
+
+            converted_kwargs = {}
+            for key, val in kwargs.items():
+                kwarg_unit = kwargs_units.get(key, None)
+                converted_kwargs[key] = _check_or_convert_to_then_strip(val, kwarg_unit)
+
+            results = func(*converted_args, **converted_kwargs)
+
+            if results is None:
+                if return_units:
+                    raise TypeError(
+                        "Expected function to return something, but function returned None"
+                    )
+            else:
+                if return_units is None:
+                    # ignore types and units of return values
+                    return results
+                else:
+                    # handle case of function returning only one result by promoting to 1-element tuple
+                    if not isinstance(results, tuple):
+                        results = (results,)
+                    if isinstance(return_units, (tuple, list)):
+                        # avoid mutating return_units because that would change the variables' scope
```

Comment:
> Just for reference: where did you notice an issue with the scope?

Specifically if you take the code as it is right now (https://github.com/xarray-contrib/pint-xarray/pull/143/commits/a5a2493df0c911c9a1211793059f68be34de3c97) and simply replace `return_units_list` -> `return_units` everywhere, attempting to import this module gives an error

```python
self = <pint_xarray.tests.test_checking.TestExpects object at 0x7fd16b5e3ac0>

    def test_single_arg(self):
        @expects("degC")
        def above_freezing(temp):
            return temp > 0
    
        f_q = pint.Quantity(20, units="degF")
>       assert not above_freezing(f_q)

pint_xarray/tests/test_checking.py:18: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (<Quantity(20, 'degree_Fahrenheit')>,), kwargs = {}, converted_args = [array(-6.66666667)], arg = <Quantity(20, 'degree_Fahrenheit')>, arg_unit = 'degC'
converted_arg = array(-6.66666667), converted_kwargs = {}, results = False

    @functools.wraps(func)
    def _unit_checking_wrapper(*args, **kwargs):
    
        # check same number of arguments were passed as expected
        if len(args) != len(args_units):
            raise TypeError(
                f"{len(args)} arguments were passed, but {len(args_units)} arguments were expected"
            )
    
        converted_args = []
        for arg, arg_unit in zip(args, args_units):
            converted_arg = _check_or_convert_to_then_strip(arg, arg_unit)
            converted_args.append(converted_arg)
    
        converted_kwargs = {}
        for key, val in kwargs.items():
            kwarg_unit = kwargs_units.get(key, None)
            converted_kwargs[key] = _check_or_convert_to_then_strip(val, kwarg_unit)
    
        results = func(*converted_args, **converted_kwargs)
    
        if results is None:
            if return_units:
                raise TypeError(
                    "Expected function to return something, but function returned None"
                )
        else:
>           if return_units is None:
E           UnboundLocalError: local variable 'return_units' referenced before assignment

pint_xarray/checking.py:118: UnboundLocalError
```

An explanation for [why this happens is in here](https://www.openbookproject.net/py4fun/decorator/decorator.html) (search `UnboundLocalError`). It's also possible to avoid this error by using the `nonlocal` keyword (see https://github.com/xarray-contrib/pint-xarray/pull/143/commits/9e92f21618211da7d059b8faa440f0a96b59a729), which may have been a neater solution?


> In any case, I'd add an `always_iterable` function

But actually because this also avoids mutating `return_units` in favour of creating a new iterable variable, it also solves the problem.